geps/gep-1742.md

```diff
@@ -301,8 +301,133 @@ Could not find any HTTP specific timeouts. PRs welcomed. 😊
+
+	// ...
+}
+
+// HTTPRouteTimeouts defines timeouts that can be configured for an HTTPRoute.
+// Timeout values are formatted like 1h/1m/1s/1ms as parsed by Golang time.ParseDuration
+// and MUST BE >= 1ms.
+type HTTPRouteTimeouts struct {
+	// Request specifies a timeout for responding to client HTTP requests,
+	// disabled by default.
+	//
+	// For example, the following rule will timeout if a client request is taking
+	// longer than 10 seconds to complete:
+	//
+	// ```
+	// rules:
+	// - timeouts:
+	//     request: 10s
+	//   backendRefs:
+	//     ...
+	// ```
+	//
+	// Support: Core
+	//
+	// +optional
+	Request *metav1.Duration `json:"request,omitempty"`
+
+	// BackendRequest specifies a timeout for an individual request from the gateway
+	// to a backend service.
+	// Default is the value of Request timeout.
+	//
+	// Support: Extended
+	//
+	// +optional
+	BackendRequest *metav1.Duration `json:"backendRequest,omitempty"`
+}
+```
+
+### YAML
+
+```yaml
+apiVersion: gateway.networking.k8s.io/v1beta1
+kind: HTTPRoute
+metadata:
+  name: timeout-example
+spec:
+  ...
+  rules:
+  - backendRefs:
+    - name: some-service
+      port: 8080
+    timeouts:
+      request: 10s
+      backendRequest: 2s
+```
+
+### Timeout values
+
+There a 2 kinds of timeouts that can be configured in an `HTTPRouteRule`:
+
+1. `timeouts.request` is the timeout for the Gateway API implementation to send a
+    response to a client HTTP request. Whether the gateway starts the timeout before
+    or after the entire client request stream has been received, is implementation dependent.
+    This field is required `Core` support.
+
+1. `timeouts.backendRequest` is a timeout for a single request from the gateway to a backend.
+    This field is optional `Extended` support. Typically used in conjuction with retry configuration,
+    if supported by an implementation.
+    Note that retry configuration will be the subject of a separate GEP (GEP-1731).
+
+```mermaid
+sequenceDiagram
+    participant C as Client
+    participant P as Proxy
+    participant U as Upstream
+    C->>P: Connection Started
+    note left of P: timeouts.request start time (min)
+    C->>P: Starts sending Request
+    C->>P: Finishes Headers
+    note left of P: timeouts.request start time (max)
+    C->>P: Finishes request
+    P->>U: Connection Started
+    note right of P: timeouts.backendRequest start time
+    P->>U: Starts sending Request
+    P->>U: Finishes request
+    P->>U: Finishes Headers
+    U->>P: Starts Response
+    U->>P: Finishes Headers
+    note right of P: timeouts.backendRequest end time
+    note left of P: timeouts.request end time
+    U->>P: Finishes Response
+    note right of P: Repeat if retry
+    P->>C: Starts Response
+    P->>C: Finishes Headers
+    P->>C: Finishes Response
+    Note right of P: Repeat if connection sharing
+    U->>C: Connection ended
+```
 
+Both timeout fields are string duration values as specified by ISO 8601 and MUST be >= 1ms.
```

Comment:
So I would set

```
timeout: PT1M
```

for a 1 minute timeout?

Reply:
Yeah, I noticed that too. It's actually the Golang `time.Duration` string format, which somewhere I read is "based on ISO 8601", or something like that. In the comment in front of the go struct I have this:
```
// Timeout values are formatted like 1h/1m/1s/1ms as parsed by Golang time.ParseDuration
```
Not sure what I should reference where it has the complete syntax documented? The Golang time.ParseDuration doc is pretty vague: https://pkg.go.dev/time#ParseDuration

Reply:
ah looks like this was already discussed, so maybe disregard my comment below